### ISSUE TITLE
Will ELMAH be ported to ASP.NET Core?

### ISSUE THREAD

**Issue by mgmccarthy** [elmah/Elmah#407]
I'm having a hard time finding information about this. I realize there is cloud-based elmah.io project (https://elmah.io/) that is has a beta out for .NET Core, but that's a different project and in no way affiliated with ELMAH (at least I think it is)...

any info you could provide would be helpful.

**Comment by atifaziz** [comment]
> I realize there is cloud-based elmah.io project…but that's a different project and in no way affiliated with ELMAH (at least I think it is)...

That's correct. It's a completely different project run by @ThomasArdal.

> Will ELMAH be ported to ASP.NET Core?

It's more of an “if & when time permits” than a definite “no”. If it depends entirely on me then I'm afraid for you that it'll be on my agenda and when I'm ready to ship my first ASP.NET Core project. If some people have already jumped on that bandwagon, well-versed in ASP.NET Core, care about backward compatibility & would like to help make it happen sooner then it could.

**Comment by ThomasArdal** [comment]
@mgmccarthy I'm Thomas, the founder of elmah.io and as @atifaziz says, it's a different project with what turned out as a bad naming choice.

I have promissed Atif to look at ELMAH for ASP.NET Core. I decided to implement the beta for elmah.io first, to find out how middleware are implemented for ASP.NET Core. I feel that the [implementation](https://github.com/elmahio/Elmah.Io.AspNetCore/blob/master/Elmah.Io.AspNetCore/ElmahIoMiddleware.cs) is in pretty good shape right now and I would love to merge the implementation back to the ELMAH project.

@atifaziz Would you mind taking a look at the implementation for Elmah.Io.AspNetCore and share if you would want some of the code. Maybe just for inspiration. I know there's an OWIN branch on ELMAH, but I guess that one were never actually ported to Core, right?

**Comment by atifaziz** [comment]
@ThomasArdal Thanks for sharing that Thomas but it's only a very small piece (posting an error) of the entire feature set and so the bulk of the work (UI, mailing, authorisation, log stores, refactoring of code & packages, testability, etc.) is still pending requires a fair amount of time.

> I know there's an OWIN branch on ELMAH, but I guess that one were never actually ported to Core, right?

Correct.

**Comment by ThomasArdal** [comment]
@atifaziz I know 😄 Should we create a new branch for core or do you want to continue the work on the owin branch?

**Comment by atifaziz** [comment]
A new branch on top of the owin one probably makes sense but before that, I think that the existing repo needs to be refactored into different ones: one for the core bits and then separate ones for the error log implementations so each can have its own issues, versions and release cycles. Then merge that structure into the OWIN version and then start the ASP.NET Core port. That's just thinking from the top of my head.

**Comment by ThomasArdal** [comment]
@atifaziz I'm not sure about the state of the OWIN branch. Would it make sense to create a 2.0 branch from master and make the split there? Then pick the changes from the owin branch that makes sense for core and put them into the 2.0 branch? Or do you believe that the owin branch is in a state where we could build 2.0 from that?

**Comment by atifaziz** [comment]
@ThomasArdal `master` represents the latest development, which is [already 2.0](https://github.com/elmah/Elmah/blob/dd67341a96406e66f107480d66fd6733fa07dab8/src/Elmah/AssemblyInfo.cs#L41-L42).

**Comment by ThomasArdal** [comment]
@atifaziz Ok. Is ELMAH on the owin branch still based on modules? Was looking into creating middleware for core instead.

**Comment by atifaziz** [comment]
@ThomasArdal ASP.NET Core borrows the ideas of middleware from OWIN. In fact, it even [supports OWIN middleware][aspnetcore-owin]. So naturally the `owin` branch isn't based on modules. Have a look at [`Owin.cs`][owin.cs] where most of the attempt was done & works to a very large extent. See the following extension methods that setup the middleware in the pipeline via `IAppBuilder`:

- [`UseElmahWeb`](https://github.com/elmah/Elmah/blob/78fca98bb9de94f01270cfff19843b40ccddd704/src/Elmah.Owin/Owin.cs#L137)
- [`UseElmahLogging`](https://github.com/elmah/Elmah/blob/78fca98bb9de94f01270cfff19843b40ccddd704/src/Elmah.Owin/Owin.cs#L146)
- [`UseElmahMailing`](https://github.com/elmah/Elmah/blob/78fca98bb9de94f01270cfff19843b40ccddd704/src/Elmah.Owin/Owin.cs#L146)

I mostly stopped working on all this when I found out that ASP.NET vNext (at the time) wasn't going to be based on OWIN. Wanted the let the dust settle down.

[aspnet-owin]: https://docs.asp.net/en/latest/fundamentals/owin.html
[owin.cs]: https://github.com/elmah/Elmah/blob/78fca98bb9de94f01270cfff19843b40ccddd704/src/Elmah.Owin/Owin.cs

**Comment by ThomasArdal** [comment]
@atifaziz Makes sense. I will have a look at porting the OWIN middleware to ASP.NET Core middleware when time allows. Thanks.

**Comment by milla** [comment]
Is the release date for elmah .net core planned :)

**Comment by atifaziz** [comment]
@milla See [my earlier comment](https://github.com/elmah/Elmah/issues/407#issuecomment-253589590):

> If it depends entirely on me then I'm afraid for you that it'll be on my agenda and when I'm ready to ship my first ASP.NET Core project. If some people have already jumped on that bandwagon, well-versed in ASP.NET Core, care about backward compatibility & would like to help make it happen sooner then it could.

**Comment by mgmccarthy** [comment]
@ThomasArdal any progress made on porting the OWIN middleware to ASP.NET Core middleware?

I'd like to jump in and help, but I'm currently using most of my free time on another open source project which I would like to use ELMAH on.

**Comment by ThomasArdal** [comment]
@mgmccarthy No :( Haven't had time to look at it either.

**Comment by atifaziz** [comment]
> the existing repo needs to be refactored into different ones: one for the core bits and then separate ones for the error log implementations so each can have its own issues, versions and release cycles.

This was done in #418.

**Comment by szalapski** [comment]
Any updates on how to use ELMAH or something similar in .NET Core?

**Comment by atifaziz** [comment]
@szalapski The update is that I'm afraid it's no different than [my initial comment](https://github.com/elmah/Elmah/issues/407#issuecomment-253589590):

>> Will ELMAH be ported to ASP.NET Core?
>
> It's more of an “if & when time permits” than a definite “no”. If it depends entirely on me then I'm afraid for you that it'll be on my agenda and when I'm ready to ship my first ASP.NET Core project. If some people have already jumped on that bandwagon, well-versed in ASP.NET Core, care about backward compatibility & would like to help make it happen sooner then it could.

I have shipped ASP.NET Core projects since and I'm beginning to feel the itch or miss ELMAH myself but the pain hasn't reached a level where I feel compelled to dive in alone.

As for something similar for .NET Core, I'm not aware of anything except [ElmahCore](https://github.com/ElmahCore/ElmahCore) by @barestan.

**Comment by frattaro** [comment]
I don't know why Microsoft doesn't pick this up. This library is the main reason I don't use .net core (System.DirectoryServices is another, but that is coming out in core 2.1).

**Comment by ofuochi** [comment]
@frattaro  This is exactly my case too. I can't afford the subscription for elmah.io :(
